### PR TITLE
Fix explicit `"types"` list being overridden by ATA files

### LINKED ISSUE #47709: [BUG][javascript] "types" compiler option in jsconfig.json doesn't work
Does this issue occur when all extensions are disabled?: Yes/No

- VS Code Version: 1.64.0-insider
- OS Version: Mac OS 12.1

Steps to Reproduce:

Copy of https://github.com/MicrosoftDocs/intellicode/issues/328.
CC @minestarks @tjx666
 
User reports "types" compiler option in jsconfig.json doesn't work
1. "When I enable this extension, you can see that I can still get the suggest from @types/node."
image
2. "But no issue in typescript project or disable this extension in js project:"
image

We believe that this issue relates more to JS/TS IntelliSense support than IntelliCode.

### PR DESCRIPTION
**Pull request by andrewbranch**
This is still a bit of a mess, but wanted to put something up to see if @sheetalkamat thinks this is headed in the right direction.

If `compilerOptions.types` is set, we do not return files from ATA as part of the root files of the project. Instead, we allow the automatic type reference directives to resolve from the global typings cache, the same way we do for non-relative imports. Making the resolution work was very easy, but invalidating the previously failed resolutions was a little tricky—there may be a more elegant solution. Comments inline.

Fixes #47709

### CODE REVIEW

File: src/compiler/resolutionCache.ts [hunk 1]
```diff
@@ -346,8 +355,28 @@ namespace ts {
             return primaryResult;
         }
 
-        function resolveTypeReferenceDirective(typeReferenceDirectiveName: string, containingFile: string | undefined, options: CompilerOptions, host: ModuleResolutionHost, redirectedReference?: ResolvedProjectReference, _containingSourceFile?: SourceFile, resolutionMode?: SourceFile["impliedNodeFormat"] | undefined): CachedResolvedTypeReferenceDirectiveWithFailedLookupLocations {
-            return ts.resolveTypeReferenceDirective(typeReferenceDirectiveName, containingFile, options, host, redirectedReference, typeReferenceDirectiveResolutionCache, resolutionMode);
+        function resolveTypeReferenceDirective(typeReferenceDirectiveName: string, containingFile: string | undefined, compilerOptions: CompilerOptions, host: ModuleResolutionHost, redirectedReference?: ResolvedProjectReference, _containingSourceFile?: SourceFile, resolutionMode?: SourceFile["impliedNodeFormat"] | undefined): CachedResolvedTypeReferenceDirectiveWithFailedLookupLocations {
```
Comment (andrewbranch) [inline src/compiler/resolutionCache.ts:358]: Will probably refactor this to share some repeated code with `resolveModuleName` before merging

File: src/compiler/resolutionCache.ts [hunk 2]
```diff
@@ -831,6 +861,11 @@ namespace ts {
             filesWithInvalidatedNonRelativeUnresolvedImports = filesMap;
         }
 
+        function invalidateTypingsFiles() {
+            hasChangedAutomaticTypeDirectiveNames = true;
```
Comment (andrewbranch) [inline src/compiler/resolutionCache.ts:865]: This is a bit of an abuse of this variable—the _names_ didn’t actually change; the resolution results did. The distinction isn’t important for consumers at the moment, so it’s a shame that the name doesn’t quite fit. I can add a different variable and corresponding method on `CompilerHost` unless there’s a better option.

File: src/server/project.ts [hunk 3]
```diff
@@ -1114,6 +1114,7 @@ namespace ts.server {
                 // If typing files changed, then only schedule project update
                 this.typingFiles = typingFiles;
                 // Invalidate files with unresolved imports
+                this.resolutionCache.invalidateTypingsFiles();
```
Comment (andrewbranch) [inline src/server/project.ts:1117]: It seems like it would be possible to tell the resolution cache exactly which typings files were added/removed. But I’m not sure it’s worth it, because `resolveTypeReferenceDirective` already short-circuits on resolutions that were previously successful.

**Comment by sheetalkamat** [comment]
Did you consider having ATA disabled if types is set and typeAcquisition enable is not set for jsconfig ?

**Comment by andrewbranch** [comment]
Yes, I think I’d like to do that as a follow-up. But I think this is still necessary since the user could still explicitly enable typeAcquisition. Unless we want to make that an error.

File: src/compiler/resolutionCache.ts [hunk 4]
```diff
@@ -281,6 +285,10 @@ namespace ts {
             return !!value && !!value.length;
         }
 
+        function isInvalidatedAutomaticTypeReferenceDirective(path: Path): boolean {
+            return hasChangedTypingsFiles && endsWith(path, inferredTypesContainingFile);
```
Comment (sheetalkamat) [inline src/compiler/resolutionCache.ts:289]: Isnt this exactly same as `hasChangedAutomaticTypeDirectiveNames`

Comment (andrewbranch) [reply src/compiler/resolutionCache.ts:289]: It’s called from `resolveNamesWithLocalCache` so we need to know that it was an automatic type reference directive (hence checking the path). If it were just `return hasChangedAutomaticTypeDirectiveNames`, all imports would appear invalidated too.

Comment (andrewbranch) [reply src/compiler/resolutionCache.ts:289]: Oh, you mean is `hasChangedTypingsFiles` the same as `hasChangedAutomaticTypeDirectiveNames`? No, `hasChangedAutomaticTypeDirectiveNames` can be true when `hasChangedTypingsFiles` is false.

Comment (sheetalkamat) [reply src/compiler/resolutionCache.ts:289]: oh i see .. you need this to ensure that file is invalidated and re-resolved.

File: src/server/project.ts [hunk 5]
```diff
@@ -404,7 +404,7 @@ namespace ts.server {
                 }
             });
 
-            return addRange(result, this.typingFiles) || ts.emptyArray;
+            return addRange(result, this.compilerOptions.types ? undefined : this.typingFiles) || ts.emptyArray;
```
Comment (sheetalkamat) [inline src/server/project.ts:407]: Did you try not adding typings file all the time as root and instead relying on module resolution and your modified type reference resolution ?

Comment (andrewbranch) [reply src/server/project.ts:407]: I don’t think they would be resolved and included automatically because nothing is referencing them. I think we’d have to convert this list into a list of automatic type reference directives, essentially. I think that approach makes sense logically, but it will be more code and produce the same outcome.

Comment (sheetalkamat) [reply src/server/project.ts:407]: i meant did you try running some real world scenarios and asserting if typingFiles are not in the program after program is construction is complete? May be node typings will not resolve but if thats the case why would we anyways do type Acquistion when types  is set

Comment (sheetalkamat) [reply src/server/project.ts:407]: > I don’t think they would be resolved and included automatically because nothing is referencing them

ATA installs and gives you list of files based on unresolved import right.. so there are places where it is imported and shoul;d work no?

File: src/compiler/resolutionCache.ts [hunk 6]
```diff
@@ -346,8 +355,28 @@ namespace ts {
...
+        function resolveTypeReferenceDirective(typeReferenceDirectiveName: string, containingFile: string | undefined, compilerOptions: CompilerOptions, host: ModuleResolutionHost, redirectedReference?: ResolvedProjectReference, _containingSourceFile?: SourceFile, resolutionMode?: SourceFile["impliedNodeFormat"] | undefined): CachedResolvedTypeReferenceDirectiveWithFailedLookupLocations {
+            const primaryResult = ts.resolveTypeReferenceDirective(typeReferenceDirectiveName, containingFile, compilerOptions, host, redirectedReference, typeReferenceDirectiveResolutionCache, resolutionMode);
+            if (!resolutionHost.getGlobalCache || primaryResult.resolvedTypeReferenceDirective) {
+                return primaryResult;
+            }
+            const globalCache = resolutionHost.getGlobalCache();
+            if (globalCache !== undefined && !isExternalModuleNameRelative(typeReferenceDirectiveName)) {
+                const { resolvedTypeReferenceDirective, failedLookupLocations } = loadTypeReferenceDirectiveFromGlobalCache(
+                    Debug.checkDefined(resolutionHost.globalCacheResolutionModuleName)(typeReferenceDirectiveName),
```
Comment (sheetalkamat) [inline src/compiler/resolutionCache.ts:366]: This change here should also handle node.. so i dont think we need to worry about including typings files at all..

File: src/server/project.ts (hunk 5 above)
Comment (andrewbranch) [reply src/server/project.ts:407]: Unresolved imports work, yes, but ATA also gets typings to include from package.json dependencies: https://github.com/microsoft/TypeScript/blob/46e53016e45133dd4901c4d3de62c1e249e131ae/src/jsTyping/jsTyping.ts#L224-L225

Comment (andrewbranch) [reply src/server/project.ts:407]: Here’s an example. This project has no imports, yet I get global completions for React just because it’s in my package.json.

image

Comment (sheetalkamat) [reply src/server/project.ts:407]: Hmm is it then that when types set to not include anything but unresolved imports in the ATA (that is in getTypingsName)

File: src/compiler/resolutionCache.ts (hunk 6 above)
Comment (sheetalkamat) [reply src/compiler/resolutionCache.ts:366]: So getTypingName to do ATA for unresolved imports + types array ?

File: src/server/project.ts (hunk 5 above)
Comment (andrewbranch) [reply src/server/project.ts:407]: Can you rephrase the question? 🤔 The example I’m showing has no jsconfig at all, by the way.

Comment (sheetalkamat) [reply src/server/project.ts:407]: Not to do packageJson lookups when types array is set ?
Will that achieve same result as this. I think it would be better to not do ATA for types if its not going to be used
Essentially discover types should ATA for types from unresolved imports only ?
https://github.com/microsoft/TypeScript/blob/ba4525202e8e647cd71b87c7711ad0c4de8348ee/src/jsTyping/jsTyping.ts#L151

Comment (andrewbranch) [reply src/server/project.ts:407]: Hm, yeah, that might work.

Comment (andrewbranch) [reply src/server/project.ts:407]: This makes my test with `"types": ["jquery"]` fail, but I think I might be ok with that. Maybe automatic type reference directives (from an explicit `types` list) _shouldn’t_ resolve from the global typings cache—since they’re triggered by a TS compiler option, they should resolve the same way whether you’re in the editor or using tsc.

In other words, unresolved imports can trigger special resolution behavior because imports first and foremost are written for use at runtime; bringing in the typings on top of them is an added bonus to help your experience at design time. Type reference directives (including `"types"`) are different because they have no runtime purpose; if you wrote them, you were intentionally using a TypeScript-only feature and you should do so completely by installing the types/dependencies you say you want to reference.

**Comment by andrewbranch** [comment]
Closing in favor of #49380
